### PR TITLE
Optimizing binomial(n, k) evaluation for integers 0 <= k <= n

### PR DESCRIPTION
**Pull request by sidhantnagpal**
#### Brief description of what is fixed or changed
As discussed on Gitter, the method for evaluation of `binomial(n, k)` for `0 <= k <= n`, where `k` and `n` are integers uses the [prime factorization](https://github.com/sympy/sympy/blob/master/sympy/functions/combinatorial/factorials.py#L815) of `C(n, k)` using `Legendre's Formula` for exponent of prime in factorial if estimated prime count is less than `k`.
The complexity of this method is `O(n lg(lg(n)) + pi(sqrt(n)) lg(n))`, where `pi(n)` is number of primes `<= n`.

Also, if `prime_count_estimate > k`, the [calculation](https://github.com/sympy/sympy/blob/master/sympy/functions/combinatorial/factorials.py#L841) is not optimal (as it involves evaluation and division of possibly 2 very large numbers - `falling factorial` and `factorial`). For cases like:
```python
>>> binomial(1000, 50)
>>> binomial(2000, 50)
```
where `prime_count_estimate > k`,` factorial(50)` is called and `falling factorial` is also calculated.

The calculation can be made only using integers in `O(min(k, n-k))` using a [loop](https://github.com/sympy/sympy/blob/master/sympy/functions/combinatorial/factorials.py#L845) similar to that used for non-integral `n` and integral `k` in the same method.
The underlying property used is `C(n, k) = (n * C(n-1, k-1)) / k`.

### CODE REVIEW

**Review by sidhantnagpal — COMMENTED** [review]
@jksuom travis build failing for a reason?
I tested it locally, all tests pass.

File: sympy/functions/combinatorial/factorials.py [hunk 1]
```diff
@@ -812,33 +812,10 @@ def _eval(self, n, k):
...
-                                exp = a + exp
-
-                            if exp > 0:
-                                result *= prime**exp
-                else:
-                    result = ff(n, k) / factorial(k)
+                d = result = n - k + 1
+                for i in range(2, k + 1):
+                    d += 1
+                    result = result // i * d + result % i * d // i
```
Comment (jksuom) [inline sympy/functions/combinatorial/factorials.py]: I this sum preferable to `result * d // i`?

Comment (sidhantnagpal) [reply sympy/functions/combinatorial/factorials.py]: Yes, I think it is best to avoid the possibly big multiplication `result * d`, which will be slow.

Comment (jksuom) [reply sympy/functions/combinatorial/factorials.py]: `d` will be relatively small though `result` will grow. I wonder if there would be significant difference in performance.

**Comment by jksuom** [comment]
There are no errors in my local tests. I suspect that the package received by travis must have been damaged in transport.

File: sympy/functions/combinatorial/factorials.py (hunk 1 above)
Comment (sidhantnagpal) [reply sympy/functions/combinatorial/factorials.py]: Yes `d` is relatively small as compared to `result` but it is still significant in comparison to `i`.
The former term is a contributing term (no big intermediate result is obtained), but the latter term uses `(i - 1) * d` as the intermediate product which will be significantly smaller than `result * d`.

**Comment by jksuom** [comment]
I have restarted one job to see if it would get a new version not buffered locally at travis site.

**Comment by sidhantnagpal** [comment]
@jksuom Thank you, I will keep a check on the status.

**Comment by sidhantnagpal** [comment]
The job failed, with same errors.

File: sympy/functions/combinatorial/factorials.py (hunk 1 above)
Comment (isuruf) [reply sympy/functions/combinatorial/factorials.py]:
```IPython
In [67]: def f(n, k):
    ...:   d = result = n-k + 1
    ...:   for i in range(2, k + 1):
    ...:     d += 1
    ...:     result = result // i * d + result % i * d // i
    ...:   return result
    ...: 

In [68]: def g(n, k):
    ...:   d = result = n-k + 1
    ...:   for i in range(2, k + 1):
    ...:     d += 1
    ...:     result = result * d // i
    ...:   return result
    ...: 

In [69]: %timeit f(1000, 500)
100 loops, best of 3: 3.09 ms per loop

In [70]: %timeit g(1000, 500)
1000 loops, best of 3: 1.51 ms per loop

In [71]: %timeit f(1000, 50)
10000 loops, best of 3: 153 µs per loop

In [72]: %timeit g(1000, 50)
10000 loops, best of 3: 71.2 µs per loop

In [73]: %timeit f(100, 50)
10000 loops, best of 3: 95.9 µs per loop

In [74]: %timeit g(100, 50)
10000 loops, best of 3: 47.1 µs per loop
```

Comment (sidhantnagpal) [reply sympy/functions/combinatorial/factorials.py]:
```
In [5]: %timeit f(21232, 42)
1000000 loops, best of 3: 736 ns per loop

In [6]: %timeit g(21232, 42)
1000000 loops, best of 3: 834 ns per loo
```

But for smaller cases, I think the more number of operations (especially modulo and division) cause the slowdown.

Comment (sidhantnagpal) [reply sympy/functions/combinatorial/factorials.py]: Interestingly,
```
In [21]: %timeit g(88932, 80)
1000000 loops, best of 3: 756 ns per loop

In [22]: %timeit f(88932, 80)
1000000 loops, best of 3: 1.27 µs per loop
```
For majority cases, `g(n, k)` is faster than `f(n, k)` - tested for around 50 cases.
Should I commit to that?

Thank you @jksuom @isuruf for the review.

**Comment by sidhantnagpal** [comment]
Benchmarks for the PR:

```
In [1]: from sympy import *

In [2]: from math import sqrt as _sqrt

In [3]: %paste
def f(n, k):
        d = result = n - k + 1
        for i in range(2, k + 1):
                d += 1
                result = result // i * d + result % i * d // i
                return result

def g(n, k):
        d = result = n - k + 1
        for i in range(2, k + 1):
                d += 1
                result = result * d // i
                return result

def h(n, k):
        from sympy.core import N
        prime_count_estimate = N(n / log(n))

        # if the number of primes less than n is less than k, use prime sieve method
        # otherwise it is more memory efficient to compute factorials explicitly
        if prime_count_estimate < k:
                M, result = int(_sqrt(n)), 1
                for prime in sieve.primerange(2, n + 1):
                        if prime > n - k:
                                result *= prime
                        elif prime > n // 2:
                                continue
                        elif prime > M:
                                if n % prime < k % prime:
                                        result *= prime
                                else:
                                        exp = a = 0

                                        while n > 0:
                                                a = int((n % prime) < (k % prime + a))
                                                n, k = n // prime, k // prime
                                                exp += a

                                        if exp > 0:
                                                result *= prime**exp
        else:
                result = ff(n, k) / factorial(k)
        return Integer(result)

## -- End pasted text --

In [4]: %timeit f(1735, 43)
1000000 loops, best of 3: 789 ns per loop

In [5]: %timeit g(1735, 43)
1000000 loops, best of 3: 620 ns per loop

In [6]: %timeit h(1735, 43)
1000 loops, best of 3: 348 µs per loop

In [7]: %timeit f(4321, 51)
1000000 loops, best of 3: 888 ns per loop

In [8]: %timeit g(4321, 51)
1000000 loops, best of 3: 719 ns per loop

In [9]: %timeit h(4321, 51)
1000 loops, best of 3: 486 µs per loop

In [10]: %timeit f(500, 51)
1000000 loops, best of 3: 913 ns per loop

In [11]: %timeit g(500, 51)
1000000 loops, best of 3: 703 ns per loop

In [12]: %timeit h(500, 51)
1000 loops, best of 3: 338 µs per loop

In [13]: %timeit f(100, 48)
1000000 loops, best of 3: 796 ns per loop

In [14]: %timeit g(100, 48)
1000000 loops, best of 3: 684 ns per loop

In [15]: %timeit h(100, 48)
1000 loops, best of 3: 350 µs per loop
```

**Comment by sidhantnagpal** [comment]
Though `f(n, k)` and `g(n, k)` have similar performance superior to `h(n, k)`, turns out that `g(n, k)` outperforms `f(n, k)` (possibly at the expense of multiple division/modulo operations in `f`) in most cases.

**Comment by jksuom** [comment]
The failure of
```
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.1.2.dev0-py2.7.egg/sympy/functions/combinatorial/tests/test_comb_factorials.py", line 291, in test_binomial
    assert binomial(10, 10) == 1
```
is particularly puzzling since this evaluation has nothing to do with the changed code. It still relies on `d = n - k` on [this line](https://github.com/sympy/sympy/blob/521484b2bc0f405a384f371ceb5dca6131c0da1b/sympy/functions/combinatorial/factorials.py#L855). At least, that is the case with my local copy of this PR. Could that be different in Travis?

**Comment by sidhantnagpal** [comment]
Yes, local copy of this PR doesn't give any errors in `test_comb_factorials.py`, `test_recurr.py`, and `test_finite_rv.py`. Could timeout be an issue?

**Comment by jksuom** [comment]
> Could timeout be an issue?

I don't think so. The errors occur persistently in the same locations.

**Comment by asmeurer** [comment]
You need to merge the branch into master to reproduce the errors (Travis does this automatically).

**Comment by jksuom** [comment]
> Travis does this automatically

I have suspected something like this but seen no do documentation on what exactly is happening.

**Comment by asmeurer** [comment]
I don't know where it's documented, but if you click the commit at the top of the build, you'll see something like https://github.com/sympy/sympy/commit/16c0936e595a35cce6025c4a0d0b9379f81ed7be, which is a merge commit with master automatically generated by GitHub.

**Comment by jksuom** [comment]
OK, now I see the error. This code

                d = result = n - k + 1
                for i in range(2, k + 1):
                    d += 1
                    result = result * d // i
                return Integer(result)

expects that k is at least 1. That is not necessarily true after `k = n - k`.

**Comment by sidhantnagpal** [comment]
Yes, valid error. How does it explain failing travis builds and passed local tests?

@jksuom Fix it by using the following?
```
d, result = n - k, 1
for i in range(1, k + 1):
     d += 1
     result = result * d // i
     return Integer(result)
```
And does this need to be documented?

**Comment by jksuom** [comment]
That looks good to me. I see no need for additional comments.

**Comment by jksuom** [comment]
> How does it explain failing travis builds and passed local tests?

You can reproduce the errors locally after merging with the current master.

**Comment by jksuom** [comment]
This looks good, thanks.

**Comment by sidhantnagpal** [comment]
@jksuom Thank you.

**Pull request merged**